src/random_ray/iteration.cpp

```diff
@@ -0,0 +1,660 @@
+    random_ray::n_source_regions, MPI_INT, MPI_MAX, mpi::intracomm);
+
+  // The position variable is more complicated to reduce than the others,
+  // as we do not want the sum of all positions in each cell, rather, we
+  // want to just pick any single valid position. Thus, we perform a gather
+  // and then pick the first valid position we find for all source regions
+  // that have had a position recorded. This operation does not need to
+  // be broadcast back to other ranks, as this value is only used for the
+  // tally conversion operation, which is only performed on the master rank.
+  // While this is expensive, it only needs to be done for active batches,
+  // and only if we have not mapped all the tallies yet. Once tallies are
+  // fully mapped, then the position vector is fully populated, so this
+  // operation can be skipped.
+
+  // First, we broadcast the fully mapped tally status variable so that
+  // all ranks are on the same page
+  int mapped_all_tallies_i = static_cast<int>(mapped_all_tallies);
+  MPI_Bcast(&mapped_all_tallies_i, 1, MPI_INT, 0, mpi::intracomm);
+
+  // Then, we perform the gather of position data, if needed
+  if (simulation::current_batch > settings::n_inactive &&
+      !mapped_all_tallies_i) {
+
+    // Master rank will gather results and pick valid positions
+    if (mpi::master) {
+      // Initialize temporary vector for receiving positions
+      std::vector<std::vector<Position>> all_position;
+      all_position.resize(mpi::n_procs);
+      for (int i = 0; i < mpi::n_procs; i++) {
+        all_position[i].resize(random_ray::n_source_regions);
+      }
+
+      // Copy master rank data into gathered vector for convenience
+      all_position[0] = random_ray::position;
+
+      // Receive all data into gather vector
+      for (int i = 1; i < mpi::n_procs; i++) {
+        MPI_Recv(all_position[i].data(), random_ray::n_source_regions * 3,
+          MPI_DOUBLE, i, 0, mpi::intracomm, MPI_STATUS_IGNORE);
+      }
+
+      // Scan through gathered data and pick first valid cell posiiton
+      for (int sr = 0; sr < random_ray::n_source_regions; sr++) {
+        if (random_ray::position_recorded[sr] == 1) {
+          for (int i = 0; i < mpi::n_procs; i++) {
+            if (all_position[i][sr].x != 0.0 || all_position[i][sr].y != 0.0 ||
+                all_position[i][sr].z != 0.0) {
+              random_ray::position[sr] = all_position[i][sr];
+              break;
+            }
+          }
+        }
+      }
+    } else {
+      // Other ranks just send in their data
+      MPI_Send(random_ray::position.data(), random_ray::n_source_regions * 3,
+        MPI_DOUBLE, 0, 0, mpi::intracomm);
+    }
+  }
+
+  // For the rest of the source region data, we simply perform an all reduce,
+  // as these values will be needed on all ranks for transport during the
+  // next iteration.
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::volume.data(),
```

Comment:
I see why you didn't pack this into a vector of structs now. Surely there is some stride argument that would make this work, right?

Reply:
As far as I can recall, you have to make a custom MPI data type to do strides which is not pretty.

The bigger issue with switching to an array of structs (AOS) is that many of the iteration functions only operate on a single field, so the current "glob of global arrays" (basically an SOA) is way more efficient for those operations, both due to the cache efficiency and the ability to vectorize. E.g.,

```C++
// Normalize Scalar flux to total distance travelled by all rays this iteration
#pragma omp parallel for
  for (int64_t e = 0; e < random_ray::scalar_flux_new.size(); e++) {
    random_ray::scalar_flux_new[e] *= normalization_factor;
  }

// Accumulate cell-wise ray length tallies collected this iteration, then
// update the simulation-averaged cell-wise volume estimates
#pragma omp parallel for
  for (int64_t sr = 0; sr < random_ray::n_source_regions; sr++) {
    random_ray::volume_t[sr] += random_ray::volume[sr];
    random_ray::volume[sr] =
      random_ray::volume_t[sr] * volume_normalization_factor;
  }
```

are much faster with SOA, and there's a bunch of operations like this throughout that get run each iteration.

That said, there's other areas (e.g., the flux attenuation kernel), where the AOS pattern may potentially have some advantages that could outweigh any extra iteration costs. It may be best left to the future optimization PR(s) (tracked with issue #2844) to experiment with this. For the time being though, I think it would make sense to at least move it from the "glob of global arrays" to a cleaner SOA pattern, which will in any event make things more organized for when statepoint/restart support is added in (tracked by #2833).